### PR TITLE
Add OnSelectedCandidatePairChange callback

### PR DESCRIPTION
**Pull request by mjmac**
Non-standard, but useful for logging the chosen candidate.

### CODE REVIEW

**Comment by mjmac** [comment]
It may be a little spammy, but since it only happens during the initial setup phase it doesn't bother me. This results in logging like the following (pions logging mixed in with app logging):

> Jan 28 12:36:22 c82a14476e150000 CameraServer[727]: ice INFO: 2019/01/28 12:36:22 Selected candidate pair: prio 24552 (local, prio 3776) host 192.168.1.5:55987 <-> relay 159.203.64.229:53015 related 107.77.224.37:45236 (remote, prio 500)
Jan 28 12:36:22 c82a14476e150000 CameraServer[727]: ice INFO: 2019/01/28 12:36:22 Setting new connection state: Connected
Jan 28 12:36:22 c82a14476e150000 CameraServer[727]: pc INFO: 2019/01/28 12:36:22 ICE candidate pair selected: (local) host 192.168.1.5:55987 <-> (remote) relay 159.203.64.229:53015 related 107.77.224.37:45236
Jan 28 12:36:22 c82a14476e150000 CameraServer[727]: pc INFO: 2019/01/28 12:36:22 ICE connection state changed: Connected
Jan 28 12:36:22 c82a14476e150000 CameraServer[727]: proxy.go:474: PeerConnection 69F7DC05-5D14-44EE-A2CA-64C2EEF57E7B: Chose ICE candidate pair: host 192.168.1.5:55987 <-> relay 159.203.64.229:53015 related 107.77.224.37:45236
Jan 28 12:36:22 c82a14476e150000 CameraServer[727]: proxy.go:458: PeerConnection 69F7DC05-5D14-44EE-A2CA-64C2EEF57E7B: New ICE connection state: Connected

**Comment by Sean-Der** [comment]
Hey @mjmac 

I think I saw some of the conversation on Slack, but my 2 cents (and hopefully just repeating what was already agreed on)

I am not against adding something non-standard that is helpful. Seems like a good opportunity to to innovate.

I am not sure, but would adding any of the ones in the spec satisfy your logging needs? Pulled from [here](https://www.w3.org/TR/webrtc/#interface-definition)
```
onicecandidate of type EventHandler

The event type of this event handler is icecandidate.

onicecandidateerror of type EventHandler
The event type of this event handler is icecandidateerror.

onsignalingstatechange of type EventHandler
The event type of this event handler is signalingstatechange.

oniceconnectionstatechange of type EventHandler
The event type of this event handler is iceconnectionstatechange

onicegatheringstatechange of type EventHandler
The event type of this event handler is icegatheringstatechange.

onconnectionstatechange of type EventHandler
The event type of this event handler is connectionstatechange.
```

**Comment by mjmac** [comment]
Hi @Sean-Der... After talking with @backkem, I changed the callback to OnCandidatePairChange to match the ORTC spec for RTCIceTransport: https://draft.ortc.org/#dom-rtcicetransport-oncandidatepairchange

I think the question is whether or not we mind exposing this via the RTCPeerConnection API even though it's not part of the official WebRTC spec. I don't see anything in the standard RTCPeerConnection callbacks that will make the selected candidate pair available to a user-supplied callback, but I'm willing to believe that I missed something... In my use case, this is an important detail to log so that we have a record of which connection method was used in order to retroactively diagnose connectivity/performance issues.

**Comment by backkem** [comment]
@mjmac The ICE Transport isn't exposed on PeerConnection because WebRTC allows for separate ICE connections to be made for each media stream. The correct way to add this to the WebRTC would be:
- `RTCIceTransport` has an [onselectedcandidatepairchange ](https://w3c.github.io/webrtc-pc/#dom-rtcicetransport-onselectedcandidatepairchange). The name diverts in ORTC, maybe because the WebRTC api doesn't provide info about the selected pair. I suggest we just keep one.
- `RTCDtlsTransport` has an [iceTransport](https://w3c.github.io/webrtc-pc/#dom-rtcdtlstransport-icetransport) attribute that can be used to get the `RTCIceTransport`.
- `RTCRtpSender`, `RTCRtpReceiver` and `RTCSctpTransport` has a `transport` attribute that can be used to get the `RTCDtlsTransport`.
- `RTCPeerConnection` has a [sctp](https://w3c.github.io/webrtc-pc/#rtcpeerconnection-interface-extensions-0) attribute that can be used to get the `RTCDtlsTransport`.
- `RTCPeerConnection` has [ways](https://w3c.github.io/webrtc-pc/#rtcpeerconnection-interface-extensions) to access the `RTCRtpSender` and `RTCRtpReceiver` objects.

**Comment by mjmac** [comment]
@backkem: OK, I'm looking at this again. I understand now that adding a single callback on PeerConnection wouldn't be sufficient.

I'm pushing another revision which removes the callback from PeerConnection but implements 
OnSelectedCandidatePairChange on ICETransport. Let me know if that's what you had in mind.

Once we're OK on that part of it, I'll need to figure out how to expose the transport to user code so that the callback can be set. I see PeerConnection has e.g. GetSenders(), but I don't see a way to get at the ICETransport for that sender unless we export something new.

**Comment by backkem** [comment]
Ok, great. I guess we can implement these:
- https://w3c.github.io/webrtc-pc/#dom-rtcrtpsender-transport
- https://w3c.github.io/webrtc-pc/#dom-rtcdtlstransport-icetransport

File: icecandidate.go [hunk 1]
```diff
@@ -141,3 +141,11 @@ func convertTypeFromICE(t ice.CandidateType) (ICECandidateType, error) {
 		return ICECandidateType(t), fmt.Errorf("unknown ICE candidate type: %s", t)
 	}
 }
+
+func (c ICECandidate) String() string {
+	ic, err := c.toICE()
+	if err != nil {
+		return fmt.Sprintf("%#v", c)
```
Comment (Sean-Der) [inline icecandidate.go]: Is this going to obscure an error. It looks like `toICE` will tell you if the IP is busted (or invalid candidate type) might be useful info.

**Comment by Sean-Der** [comment]
@backkem @mjmac should we move `pkg/ice` -> `internal/ice`? Just to make sure we don't leak anything that is non-standard into the public API?

**Comment by backkem** [comment]
Fine by me.

**Pull request merged**